runtime/commonMain/src/kotlinx/serialization/modules/SerialModuleBuilders.kt

```diff
@@ -141,11 +140,45 @@ public class SerializersModuleBuilder internal constructor(@JvmField internal va
         vararg baseClasses: KClass<*>,
         buildAction: PolymorphicModuleBuilder<Any>.() -> Unit = {}
     ) {
-        val builder = PolymorphicModuleBuilder(baseClass as KClass<Any>, null)
+        val builder = PolymorphicModuleBuilder(baseClass as KClass<Any>)
         builder.buildAction()
-        builder.buildTo(impl)
+        builder.buildTo(this)
         for (base in baseClasses) {
-            builder.changeBase(base as KClass<Any>, null).buildTo(impl)
+            builder.changeBase(base as KClass<Any>, null).buildTo(this)
         }
     }
-}
\ No newline at end of file
+
+    internal fun <T : Any> registerSerializer(
+        forClass: KClass<T>,
+        serializer: KSerializer<T>,
+        allowOverwrite: Boolean = false
+    ) {
+        if (!allowOverwrite && forClass in class2Serializer) throw SerializerAlreadyRegisteredException(forClass)
+        class2Serializer[forClass] = serializer
+    }
+
+    internal fun <Base : Any, Sub : Base> registerPolymorphicSerializer(
+        baseClass: KClass<Base>,
+        concreteClass: KClass<Sub>,
+        concreteSerializer: KSerializer<Sub>,
+        allowOverwrite: Boolean = false
+    ) {
+        val name = concreteSerializer.descriptor.serialName
+        val baseClassSerializers = polyBase2Serializers.getOrPut(baseClass, ::hashMapOf)
+        if (!allowOverwrite && concreteClass in baseClassSerializers) throw SerializerAlreadyRegisteredException(
+            baseClass,
+            concreteClass
+        )
+        baseClassSerializers[concreteClass] = concreteSerializer
+
+        val names = polyBase2NamedSerializers.getOrPut(baseClass, ::hashMapOf)
+        val previous = names.put(name, concreteSerializer)
+        if (previous != null) {
+            val conflictingClass = polyBase2Serializers[baseClass]!!.filter { it.value === previous }.keys.firstOrNull()
+            throw IllegalArgumentException("Multiple polymorphic serializers for base class '$baseClass' " +
```

Comment:
I'd suggest moving this check and exception throw a bit lines up – before actual writing to a map. By that, the module builder will be in consistent state even if someone catches this exception